File: 2025/07 July/01 Substrings of length k with k-1 distinct elements/main.py
# Link: https://www.geeksforgeeks.org/problems/substrings-of-length-k-with-k-1-distinct-elements/1


class Solution:
    def substrCount(self, s, k):
        if k > len(s):
            return 0

        n = len(s)
        cnt = [0] * 26
        ans = 0

        for i in range(k - 1):
            cnt[ord(s[i]) - ord("a")] += 1

        for i in range(k - 1, n):

            cnt[ord(s[i]) - ord("a")] += 1

            distinctCnt = sum(1 for x in cnt if x > 0)
            if distinctCnt == k - 1:
                ans += 1

            cnt[ord(s[i - k + 1]) - ord("a")] -= 1

        return ans


if __name__ == "__main__":
    sol = Solution()
    s = "abcc"
    k = 2
    result = sol.substrCount(s, k)
    print(f"Number of substrings of length {k} with {k-1} distinct elements: {result}")

    s = "aabab"
    k = 3
    result = sol.substrCount(s, k)
    print(f"Number of substrings of length {k} with {k-1} distinct elements: {result}")

File: 2025/07 July/01 Substrings of length k with k-1 distinct elements/main.cpp
// Link: https://www.geeksforgeeks.org/problems/substrings-of-length-k-with-k-1-distinct-elements/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int substrCount(string &s, int k)
    {
        if (k > s.length())
            return 0;

        int n = s.length();

        vector<int> cnt(26, 0);
        int ans = 0;
        int distinctCnt = 0;

        for (int i = 0; i < k - 1; i++)
        {
            cnt[s[i] - 'a']++;

            if (cnt[s[i] - 'a'] == 1)
                distinctCnt++;
        }

        for (int i = k - 1; i < n; i++)
        {
            cnt[s[i] - 'a']++;
            if (cnt[s[i] - 'a'] == 1)
                distinctCnt++;

            if (distinctCnt == k - 1)
                ans++;

            cnt[s[i - k + 1] - 'a']--;

            if (cnt[s[i - k + 1] - 'a'] == 0)
                distinctCnt--;
        }

        return ans;
    }
};

int main()
{
    Solution sol;
    string s = "abcc";
    int k = 2;
    int result = sol.substrCount(s, k);
    cout << "Number of substrings of length " << k << " with " << k - 1 << " distinct elements: " << result << endl;

    s = "aabab";
    k = 3;
    result = sol.substrCount(s, k);
    cout << "Number of substrings of length " << k << " with " << k - 1 << " distinct elements: " << result << endl;
    return 0;
}